src/tools/rbd_mirror/PoolReplayer.cc

```diff
@@ -551,25 +525,228 @@ void PoolReplayer<I>::run()
-    if ((m_local_pool_watcher && m_local_pool_watcher->is_blacklisted()) ||
-	(m_remote_pool_watcher && m_remote_pool_watcher->is_blacklisted())) {
+
+    if (m_default_namespace_replayer->is_blacklisted()) {
       m_blacklisted = true;
       m_stopping = true;
+    }
+
+    for (auto &it : m_namespace_replayers) {
+      if (it.second->is_blacklisted()) {
+        m_blacklisted = true;
+        m_stopping = true;
+        break;
+      }
+    }
+
+    update_namespace_replayers();
+
+    if (m_stopping) {
+      // run it again to ensure all namespace replayers are shut down
+      update_namespace_replayers();
       break;
     }
 
-    if (!m_stopping) {
-      m_cond.wait_for(locker, 1s);
+    auto cct = reinterpret_cast<CephContext *>(m_local_io_ctx.cct());
+    auto seconds = cct->_conf.get_val<uint64_t>(
+        "rbd_mirror_pool_replayers_refresh_interval");
+    m_cond.wait_for(locker, ceph::make_timespan(seconds));
+  }
+
+  delete m_asok_hook;
+  m_asok_hook = nullptr;
+}
+
+template <typename I>
+void PoolReplayer<I>::update_namespace_replayers() {
+  dout(20) << dendl;
+
+  ceph_assert(ceph_mutex_is_locked(m_lock));
+
+  lock_leader();
+
+  std::set<std::string> mirroring_namespaces;
+  if (!m_stopping) {
+    int r = list_mirroring_namespaces(&mirroring_namespaces);
+    if (r < 0) {
+      return;
+    }
+  }
+
+  auto cct = reinterpret_cast<CephContext *>(m_local_io_ctx.cct());
+  C_SaferCond cond;
+  auto gather_ctx = new C_Gather(cct, &cond);
+  for (auto it = m_namespace_replayers.begin();
+       it != m_namespace_replayers.end(); ) {
+    auto iter = mirroring_namespaces.find(it->first);
+    if (iter == mirroring_namespaces.end()) {
+      auto namespace_replayer = it->second;
+      auto on_shut_down = new FunctionContext(
+        [this, namespace_replayer, ctx=gather_ctx->new_sub()](int r) {
+          delete namespace_replayer;
+          ctx->complete(r);
+        });
+      namespace_replayer->shut_down(on_shut_down);
+      it = m_namespace_replayers.erase(it);
+    } else {
+      mirroring_namespaces.erase(iter);
+      it++;
     }
   }
 
-  m_instance_replayer->stop();
+  for (auto &name : mirroring_namespaces) {
+    auto namespace_replayer = NamespaceReplayer<I>::create(
+        name, m_local_io_ctx, m_remote_io_ctx, m_local_mirror_uuid, m_peer.uuid,
+        m_threads, m_image_sync_throttler.get(), m_service_daemon,
+        m_cache_manager_handler);
+    auto on_init = new FunctionContext(
+        [this, namespace_replayer, name, &mirroring_namespaces,
+         ctx=gather_ctx->new_sub()](int r) {
+          if (r < 0) {
+            derr << "failed to initialize namespace replayer for namespace "
+                 << name << ": " << cpp_strerror(r) << dendl;
+            delete namespace_replayer;
+            mirroring_namespaces.erase(name);
+          } else {
+            std::lock_guard locker{m_lock};
+            m_namespace_replayers[name] = namespace_replayer;
```

Comment:
This would potentially need to tell it that it's the leader, correct?

Reply:
Yes. Below if it is the leader, it calls `handle_acquire_leader` for the newly added namespace replayers.